keps/sig-scheduling/4671-gang-scheduling/README.md

```diff
@@ -0,0 +1,927 @@
+
+### Naming
+
+* `Workload` is the resource Kind.
+* `scheduling` is the ApiGroup.
+* `spec.workload` is the name of the new field in pod.
+* Within a Workload there is a list of groups of pods. Each group represents a top-level division of pods within a Workload.  Each group can be independently gang scheduled (or not use gang scheduling). This group is named
+  <<[UNRESOLVED community feedback requested]>> `PodGroup` or `GangGroup` for the top level. <<[/UNRESOLVED]>>.
+* In a future , we expect that this group can optionally specify further subdivision into sub groups.  Each sub-group can have an index.  The indexes go from 0 to N, without repeats or gaps. These subgroups are called
+  <<[UNRESOLVED depending on previous unresolved item]>> `PodSubGroup` if `PodGroup` is chosen, or else `RankedGroup` if `GangGroup` is chosen<<[/UNRESOLVED]>>.
+* In subsequent KEPs, we expect that a sub-group can optionally specify further subdivision into pod equivalence classes.  All pods in a pod equivalence class have the same values for all fields that affect scheduling feasibility.  These pod equivalence classes are called
+  <<[UNRESOLVED depending on a previous unresolved item]>> `PodSet` if `PodGroup` is chosen, or else `EqGroup` if `GangGroup` is chosen<<[/UNRESOLVED]>>.
+
+### Associating Pod into PodGroups
+
+When a `Workload` consists of a single group of pods needing Gang Scheduling, it is clear which pods belong to the group from the `spec.workload.name` field of the pod.  However `Workload` supports listing multiple list items, and a list item can represent a single group, or a set of identical replica groups.
+In these cases, there needs to be additional information to indicate which group a pod belongs to.
+
+<<[UNRESOLVED @erictune @wojtek-t]>> 
+Two options are being considered:
+  * 
+
+ <<[/UNRESOLVED]>>
+The example below uses  `PodGroupSelector` on each `PodGroup` to identify pods.  For the Job example above, this looks like:
+
+```yaml
+apiVersion: scheduling/v1alpha1   
+kind: Workload
+metadata:
+  name: jobset
+spec:
+  podGroups:   # or gangGroups -- TBD
+    - name: "job-1"
+      gangMode: Single
+      gangSchedulingPolicy:
+        minCount: 100
+        schedulingTimeoutSeconds: 60
+      podGroupSelector:
+        jobset.sigs.k8s.io/replicatedjob-name: job-1
+    - name: "job-2"
+      gangMode: Single
+      gangSchedulingPolicy:
+        minCount: 100
+        schedulingTimeoutSeconds: 60
+      podGroupSelector:
+        jobset.sigs.k8s.io/replicatedjob-name: job-2
+
+```
+This works well for the simplest usecases, but becomes more complex with replicated gangs.
+As an example, for `LeaderWorkerSet`, while pods have `leaderworkerset.sigs.k8s.io/group-index`
+label, the actual podGroupSelector would become:
+```
+      podGroupSelector:
+        leaderworkerset.sigs.k8s.io/group-index: <replica_number>
+```
+
+Another option being considered is to specify the exact `PodGroup` and its replica number in the Pod
```

Comment:
I prefer this because the first one would require some additional complexity like how we should tell to users if pod-a is specifying workload-a, but workload-a doesn't have a group that has matching selector.

Reply:
It is aligned with others preferences (including mine) - so let me already write is as a leading option and just mention the other as an alternative.